manager/drivers/secrets.go

```diff
@@ -0,0 +1,52 @@
+package drivers
+
+import (
+	"fmt"
+	"github.com/docker/docker/pkg/plugingetter"
+	"github.com/docker/swarmkit/api"
+)
+
+const (
+	// SecretsProviderAPI is the endpoint for fetching secrets from plugins
+	SecretsProviderAPI = "/SecretProvider.GetSecret"
+
+	// SecretsProviderCapability is the secrets provider plugin capability identification
+	SecretsProviderCapability = "secretprovider"
+)
+
+// SecretDriver provides secrets from different stores
+type SecretDriver struct {
+	plugin plugingetter.CompatPlugin
+}
+
+// NewSecretDriver creates a new driver that provides third party secrets
+func NewSecretDriver(plugin plugingetter.CompatPlugin) *SecretDriver {
+	return &SecretDriver{plugin: plugin}
+}
+
+// Get gets a secret from the secret provider
+func (d *SecretDriver) Get(spec *api.SecretSpec, serviceSpec api.ServiceSpec) ([]byte, error) {
+	secretReq := &SecretsProviderRequest{Name: spec.Annotations.Name, ServiceSpec: serviceSpec}
+	var secretResp SecretsProviderResponse
+	err := d.plugin.Client().Call(SecretsProviderAPI, secretReq, &secretResp)
+	if err != nil {
+		return nil, err
+	}
+	if secretResp.Err != "" {
+		return nil, fmt.Errorf(secretResp.Err)
+	}
+	// Assign the secret
+	return []byte(secretResp.Value), nil
+}
+
+// SecretsProviderRequest is the request specification for retrieving secrets from plugins.
+type SecretsProviderRequest struct {
+	Name        string          `json:"name"`        // Name is the name of the secret plugin
+	ServiceSpec api.ServiceSpec `json:"serviceSpec"` // ServiceSpec is the task's service specification
```

Comment:
Unfortunately I don't think this will work as expected. There are some interfaces inside `api.ServiceSpec` (to support `oneof`) which wouldn't be marshalled to JSON correctly.

We have code inside Docker itself that converts between this gRPC format and another representation meant for marshalling to JSON: https://github.com/moby/moby/blob/master/daemon/cluster/convert/service.go

I guess we should vendor this code and use it to serialize to JSON. It would be a little bit weird, because the JSON format has so far been outside the scope of swarmkit, but as long as we have a JSON format in use somewhere else, we might as well be consistent with it. Also, vendoring moby code into swarmkit even though swarmkit is imported by moby would be unfortunate - but sort of unavoidable, since the conversion code depends on the Docker API definitions, which live in moby. I hesitated to suggest this, but I can't think of a better solution. I don't like the idea of copying the conversion types/code because I think they will drift out of sync over time.

Note that swarmkit already vendors this repository as `github.com/docker/docker` (don't ask...).

In the longer term, we could look at moving the swarmkit-specific parts of the JSON API and the conversion code to the swarmkit repo to avoid cyclic vendoring, but I'm not completely sure whether this makes sense and whether there are obstacles. @thaJeztah: Who are the right people to talk to about this?

I would suggest, however, that we redact the `RegistryAuth` field before sending a service spec to a plugin. This is potentially sensitive information.